pkg/controllers/sync/controller.go

```diff
@@ -1104,58 +1104,113 @@ func (s *SyncController) reconcileClusterForCascadingDeletion(
 		return worker.StatusAllOK
 	}
 
-	// cascading-delete is enabled, wait for member objects to be deleted
+	if !clusterutil.IsClusterReady(&cluster.Status) && cascadingdeletion.IsCascadingDeleteEnabled(cluster) {
+		return worker.Result{RequeueAfter: &s.cascadingDeletionRecheckDelay}
+	}
+
+	// cascading-delete is enabled, delete the member objects managed by admiral
 	ftcLister := s.ftcManager.GetFederatedTypeConfigLister()
 	ftcs, err := ftcLister.List(labels.Everything())
 	if err != nil {
-		logger.Error(err, "failed to get ftc lister")
+		logger.Error(err, "Failed to get ftc lister")
 		return worker.StatusError
 	}
 
+	// concurrent operations for each ftc
+	wg := sync.WaitGroup{}
 	remainingByGVK := make(map[string]string, len(ftcs))
-	for _, ftc := range ftcs {
-		gvk := ftc.GetSourceTypeGVK().String()
-		resourceLister, hasSynced, exists := s.fedInformerManager.GetResourceLister(
-			ftc.GetSourceTypeGVK(),
-			cluster.Name,
-		)
-		if !exists {
-			remainingByGVK[gvk] = fmt.Sprintf("failed to get resource lister for %s", gvk)
-			continue
-		}
+	for i := range ftcs {
+		wg.Add(1)
+		go func(ftc *fedcorev1a1.FederatedTypeConfig) {
+			defer wg.Done()
+			gvk := ftc.GetSourceTypeGVK().String()
+
+			// get all managed cluster objects according to ftc type
+			clusterObjs, namespaces, err := s.listManagedClusterObjects(ctx, ftc, cluster.Name)
+			if err != nil {
+				logger.Error(err, "Failed to list clusterObjects")
+				remainingByGVK[gvk] = fmt.Sprintf("failed to list cluster objects for %s", gvk)
```

Comment:
A risk of concurrent map writes

Reply:
If the gvk of each ftc is different, each thread writes to different keys in the map, and there should be no conflicts.